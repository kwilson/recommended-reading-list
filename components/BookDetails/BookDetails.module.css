.container {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: auto;
  gap: 1rem;
  grid-template-areas:
    'title'
    'author'
    'isbn'
    'image'
    'otherBooks';
}

@media screen and (min-width: 760px) {
  .container {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: min-content min-content max-content min-content;
    gap: 1rem;
    grid-template-areas:
      'title title'
      'author author'
      'image isbn'
      'image otherBooks';
  }
}

.title {
  grid-area: title;
  margin-bottom: 0;
}
.image {
  grid-area: image;
}
.image img {
  border: solid 1px var(--border-color);
  height: auto;
  max-width: 100%;
}
.author {
  grid-area: author;
}
.isbn {
  grid-area: isbn;
  align-self: start;
}
.otherBooks {
  grid-area: otherBooks;
  align-self: start;
}
