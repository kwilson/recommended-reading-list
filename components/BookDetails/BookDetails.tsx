import { FunctionComponent, useEffect, useState } from 'react';
import { IBook } from '../../data/books';
import { getBooksByAuthor } from '../../data/books';
import { LoaderContainer } from '../LoaderContainer';
import { BookList } from '../BookList';

import styles from './BookDetails.module.css';

interface IProps {
  id: string;
  title: string;
  author: string;
  isbn: string | null;
  cover: string;
}

export const BookDetails: FunctionComponent<IProps> = ({
  id,
  title,
  author,
  isbn,
  cover,
}) => {
  const [relatedBooks, setRelatedBooks] = useState<IBook[] | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    setIsLoading(true);
    getBooksByAuthor(author, [id])
      .then(setRelatedBooks)
      .then(() => setIsLoading(false));
  }, [author, id]);

  return (
    <div className={styles.container} data-testid="book-details">
      <h1 className={styles.title} data-testid="title">
        {title}
      </h1>
      <div className={styles.author}>
        by <span data-testid="author">{author}</span>
      </div>
      <div className={styles.isbn}>
        ISBN: <span data-testid="isbn">{isbn ?? 'not available'}</span>
      </div>

      <div className={styles.image}>
        <img
          data-testid="image"
          className={styles.image}
          src={cover}
          alt={`${title} book cover`}
        />
      </div>

      <aside className={styles.otherBooks}>
        <h2>Other books by this author</h2>
        <LoaderContainer loading={isLoading}>
          {relatedBooks && <BookList books={relatedBooks} />}
        </LoaderContainer>
      </aside>
    </div>
  );
};
