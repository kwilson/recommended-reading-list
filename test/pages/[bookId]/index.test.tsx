import { render, screen, waitFor } from '@testing-library/react';
import { useRouter } from 'next/router';

import {
  getBook,
  getBooksByAuthor,
  IBook,
  IBookDetails,
} from '../../../data/books';
import BookDetails from '../../../pages/[bookId]';

jest.mock('../../../data/books');
const mockGetBook = getBook as jest.Mock<ReturnType<typeof getBook>>;
const mockGetBooksByAuthor = getBooksByAuthor as jest.Mock<
  ReturnType<typeof getBooksByAuthor>
>;

jest.mock('next/router');
const mockRouter = useRouter as jest.Mock<ReturnType<typeof useRouter>>;

describe(`<${BookDetails.name} />`, () => {
  it('renders as expected when loading', async () => {
    const promise = new Promise<IBookDetails>((resolve) => {
      setTimeout(resolve, 1000);
    });
    mockGetBook.mockReturnValueOnce(promise);

    const { asFragment } = render(<BookDetails />);

    await waitFor(() => screen.getByTestId('loader'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as expected when loaded', async () => {
    const book: IBookDetails = {
      id: 'id value',
      title: 'title value',
      author: 'author value',
      cover: 'cover value',
      isbn: null,
      subtitle: 'subtitle value',
      published: 'published value',
      publisher: null,
      pages: null,
      description: null,
      website: null,
    };

    mockRouter.mockReturnValueOnce({
      query: { bookId: '1' },
    } as any);

    mockGetBook.mockResolvedValueOnce(book);

    const promise = new Promise<IBook[]>((resolve) => {
      setTimeout(resolve, 1000);
    });
    mockGetBooksByAuthor.mockReturnValueOnce(promise);

    const { asFragment } = render(<BookDetails />);

    await waitFor(() => screen.getByTestId('book-details'));
    await waitFor(() => screen.getByText(book.title));
    expect(asFragment()).toMatchSnapshot();
  });

  it.only('redirects if the book fails to load', async () => {
    mockRouter.mockReturnValueOnce({
      query: { bookId: '1' },
    } as any);

    mockGetBook.mockResolvedValueOnce(null);

    const promise = new Promise<IBook[]>((resolve) => {
      setTimeout(resolve, 1000);
    });
    mockGetBooksByAuthor.mockReturnValueOnce(promise);

    const { asFragment } = render(<BookDetails />);

    await waitFor(() => screen.getByTestId('not-found'));

    expect(asFragment()).toMatchSnapshot();
  });
});
