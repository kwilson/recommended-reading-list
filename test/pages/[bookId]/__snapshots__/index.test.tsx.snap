// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookDetails /> redirects if the book fails to load 1`] = `
<DocumentFragment>
  <div
    data-testid="not-found"
  >
    <h1>
      Not Found
    </h1>
    <p>
      Sorry, the book you are trying to view has not been found.
    </p>
    <p>
      <a
        href="/"
      >
        Back to book list
      </a>
    </p>
  </div>
</DocumentFragment>
`;
